### PR TITLE
core: Walk regions for rewritings instead of ops

### PR DESCRIPTION
**Pull request by math-fehr**
Currently, the pattern rewriter walker walks an operation, and tries to apply patterns on all contained operations, including itself. This is not what we would want, as rewriting a toplevel operation is dangerous, if that operation does not have a parent.

This commit changes the walker to walk regions instead of operations, so it is clear that the toplevel operation will not be modified. This is similar to what MLIR currently does in applyPatternsAndFoldGreedily.

### CODE REVIEW

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 1]
```diff
@@ -377,17 +377,21 @@ def test_insert_op_at_start():
     """Test rewrites where operations are inserted with a given position."""
 
     prog = """"builtin.module"() ({
-  %0 = "arith.constant"() <{"value" = 5 : i32}> : () -> i32
+  "test.op"() ({
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:380]: Why did this need to change?

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 2]
```diff
@@ -377,17 +377,21 @@ def test_insert_op_at_start():
...
+    %0 = "arith.constant"() <{"value" = 42 : i32}> : () -> i32
+    %1 = "arith.constant"() <{"value" = 5 : i32}> : () -> i32
+  }) : () -> ()
 }) : () -> ()"""
 
     class Rewrite(RewritePattern):
         @op_type_rewrite_pattern
-        def match_and_rewrite(self, mod: ModuleOp, rewriter: PatternRewriter):
+        def match_and_rewrite(self, mod: test.TestOp, rewriter: PatternRewriter):
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:394]:
```suggestion
        def match_and_rewrite(self, op: test.TestOp, rewriter: PatternRewriter):
```

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 3]
```diff
@@ -404,20 +408,24 @@ def test_insert_op_before():
...
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:425]:
```suggestion
        def match_and_rewrite(self, op: test.TestOp, rewriter: PatternRewriter):
```

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 4]
```diff
@@ -377,17 +377,21 @@ def test_insert_op_at_start():
...
             new_cst = Constant.from_int_and_width(42, i32)
 
             rewriter.insert_op(new_cst, InsertPoint.at_start(mod.regions[0].blocks[0]))
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:397]:
```suggestion
            rewriter.insert_op(new_cst, InsertPoint.at_start(op.regions[0].blocks[0]))
```

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 5]
```diff
@@ -404,20 +408,24 @@ def test_insert_op_before():
...
 }) : () -> ()"""
 
     class Rewrite(RewritePattern):
         @op_type_rewrite_pattern
-        def match_and_rewrite(self, mod: ModuleOp, rewriter: PatternRewriter):
+        def match_and_rewrite(self, mod: test.TestOp, rewriter: PatternRewriter):
             new_cst = Constant.from_int_and_width(42, i32)
 
-            first_op = mod.ops.first
+            first_op = mod.regions[0].block.ops.first
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:428]:
```suggestion
            first_op = op.regions[0].block.ops.first
```

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 6]
```diff
@@ -433,20 +441,24 @@ def test_insert_op_after():
...
+  "test.op"() ({
+    %0 = "arith.constant"() <{"value" = 5 : i32}> : () -> i32
+    %1 = "arith.constant"() <{"value" = 42 : i32}> : () -> i32
+  }) : () -> ()
 }) : () -> ()"""
 
     class Rewrite(RewritePattern):
         @op_type_rewrite_pattern
-        def match_and_rewrite(self, mod: ModuleOp, rewriter: PatternRewriter):
+        def match_and_rewrite(self, mod: test.TestOp, rewriter: PatternRewriter):
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:458]:
```suggestion
        def match_and_rewrite(self, op: test.TestOp, rewriter: PatternRewriter):
```

File: tests/pattern_rewriter/test_pattern_rewriter.py [hunk 7]
```diff
@@ -433,20 +441,24 @@ def test_insert_op_after():
...
             new_cst = Constant.from_int_and_width(42, i32)
 
-            first_op = mod.ops.first
+            first_op = mod.regions[0].block.ops.first
```
Comment (alexarice) [inline tests/pattern_rewriter/test_pattern_rewriter.py:461]:
```suggestion
            first_op = op.regions[0].block.ops.first
```

File: xdsl/pattern_rewriter.py [hunk 8]
```diff
@@ -773,34 +773,34 @@ def rewrite_module(self, module: ModuleOp) -> bool:
...
-                op_was_modified |= self.post_walk_func(op, pattern_listener)
+                op_was_modified |= self.post_walk_func(region, pattern_listener)
 
         return result
 
-    def _populate_worklist(self, op: Operation) -> None:
+    def _populate_worklist(self, op: Operation | Region | Block) -> None:
         """Populate the worklist with all nested operations."""
         # We walk in reverse order since we use a stack for our worklist.
         for sub_op in op.walk(
```
Comment (alexarice) [inline xdsl/pattern_rewriter.py:806]: Is `walk` defined for Operations, Regions, and Blocks?

**Review by superlopuh — APPROVED** [review]

File: xdsl/transforms/dead_code_elimination.py [hunk 9]
```diff
@@ -159,14 +159,12 @@ def region_dce(region: Region, listener: PatternRewriterListener | None = None)
     return live_set.changed
 
 
-def op_dce(op: Operation, listener: PatternRewriterListener | None = None):
-    changed = tuple(region_dce(region, listener) for region in op.regions)
-
-    return any(changed)
+def op_dce(region: Region, listener: PatternRewriterListener | None = None):
```
Comment (alexarice) [inline xdsl/transforms/dead_code_elimination.py:162]: Maybe just scrap this function?

**Review by alexarice — APPROVED** [review]
Left a few small comments but looks good to me.

PS. why don't we have the batched suggestion thing enabled

File: xdsl/pattern_rewriter.py (hunk 8 above)
Comment (math-fehr) [reply xdsl/pattern_rewriter.py:806]: Yep, and pyright is happy with it!

**Pull request merged**